toolchain/common/enum_base.h

```diff
@@ -0,0 +1,179 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_COMMON_ENUM_BASE_H_
+#define CARBON_TOOLCHAIN_COMMON_ENUM_BASE_H_
+
+#include <cstdint>
+#include <type_traits>
+
+#include "common/ostream.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace Carbon::Internal {
+
+// CRTP-style base class used to define the common pattern of Carbon enum-like
+// classes. The result is a class with named constants similar to enumerators,
+// but that are normal classes, can contain other methods, and support a `name`
+// method and printing the enums. These even work in switch statements and
+// support `case MyEnum::Name:`.
+//
+// It is specifically designed to compose with X-MACRO style `.def` files that
+// stamp out all the enumerators.
+//
+// It also supports some opt-in APIs that classes can enable by `using` the
+// names to make them public: `AsInt` and `FromInt` allow converting to and from
+// the underlying type of the enumerator.
+//
+// Users must be in the `Carbon` namespace and should look like the following.
```

Comment:
```suggestion
// Users must be in the `Carbon` namespace and should look like the following.
//
```

Suggesting a little whitespace to separate file content.